pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,251 @@
+		configMapListerForTargetNamespace: kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Lister().ConfigMaps(operatorclient.TargetNamespace),
+		hasNetworkInformerSyncedFn:        hasNetworkInformerSyncedFn,
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
+		return nil
+	}
+	// stop if the machine API is not functional
+	if isFunctional, err := c.isMachineAPIFunctionalFn(); err != nil {
+		return err
+	} else if !isFunctional {
+		return nil
+	}
+	etcdEndpointsConfigMap, err := c.configMapListerForTargetNamespace.Get("etcd-endpoints")
+	if err != nil {
+		return err // should not happen
+	}
+	for len(etcdEndpointsConfigMap.Data) == 0 {
+		return nil
+	}
+	for _, potentialMemberToRemoveIP := range etcdEndpointsConfigMap.Data {
+		memberLocator := potentialMemberToRemoveIP
+		machine, err := c.getMachineForMember(potentialMemberToRemoveIP)
+		if err != nil {
+			klog.Errorf("unable to get a machine for member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if machine != nil {
+			klog.V(4).Infof("cannot remove member: %v because a machine resource still exists (%v/%v)", memberLocator, machine.Name, machine.UID)
+			continue
+		}
+
+		node, err := c.getNodeForMember(potentialMemberToRemoveIP)
+		if err != nil {
+			klog.Errorf("unable to get a node for member: %v, err: %v", memberLocator, err)
+			continue
+		}
+		if node != nil {
+			klog.V(4).Infof("cannot remove member: %v because a node resource still exists (%v/%v)", memberLocator, node.Name, node.UID)
+			continue
+		}
+
+		// it looks like we have found a member that isn't backed by a machine nor a node
+		// issue a live request to the etcd cluster and remove it from the cluster
+		liveMemberList, err := c.etcdClient.MemberList(ctx)
+		if err != nil {
+			return err
+		}
+		for _, liveMember := range liveMemberList {
+			liveMemberIP, err := memberToNodeInternalIP(liveMember)
+			if err != nil {
+				return err
+			}
+			if liveMemberIP != potentialMemberToRemoveIP {
+				continue
+			}
+			memberLocator = fmt.Sprintf("[ url: %v, name: %v, id: %v ]", liveMemberIP, liveMember.Name, liveMember.ID)
+			if len(liveMember.Name) == 0 {
+				// TODO: what if the member hasn't been started and the name field is empty ?!
+				klog.Errorf("unable to remove member: %v because it doesn't have a name (has the member been started?)", memberLocator)
+				break
+			}
+			err = c.etcdClient.MemberRemove(ctx, liveMember.Name)
```

Comment:
`c.etcdClient.MemberRemove()` is just our wrapper around the etcd client which actually just uses the `member.ID` field to remove members.
https://github.com/openshift/cluster-etcd-operator/blob/39e43fd6ecc4bd82705566d9da5af990dabd56ae/pkg/etcdcli/etcdcli.go#L275

I got around this in my PR by adding a similar method `MemberRemoveID()` to `pkg/etcdcli/etcdcli.go` which just uses the ID so we can still remove unstarted members.
https://github.com/hasbro17/cluster-etcd-operator/blob/62ce76b24d48c749c16f0503111bd4752149a69b/pkg/etcdcli/etcdcli.go#L340-L386

So you could do something similar and can consolidate them later to avoid refactoring the bootstrap teardown controller's [use of MemberRemove()](https://github.com/openshift/cluster-etcd-operator/blob/39e43fd6ecc4bd82705566d9da5af990dabd56ae/pkg/operator/bootstrapteardown/bootstrap_teardown_controller.go#L135) for now.

Reply:
Would it be okay to leave it as it is and open a new pr that would change the signature of this method to accept an ID instead of a Name? I could open such a PR.

Reply:
ptal - https://github.com/openshift/cluster-etcd-operator/pull/741 (changes the signature of the MemberRemove method)